src/NodeVisitor/LoopNodeVisitor.php

```diff
@@ -0,0 +1,109 @@
+<?php
+
+/*
+ * This file is part of Twig.
+ *
+ * (c) Fabien Potencier
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Twig\NodeVisitor;
+
+use Twig\Environment;
+use Twig\Error\SyntaxError;
+use Twig\Node\BlockNode;
+use Twig\Node\BreakNode;
+use Twig\Node\ForNode;
+use Twig\Node\MacroNode;
+use Twig\Node\Node;
+use Twig\Node\WithNode;
+
+final class LoopNodeVisitor implements NodeVisitorInterface
+{
+    /**
+     * @var int
+     */
+    private $depth = 0;
+
+    /**
+     * @var int[]
+     */
+    private $structureStack = [];
+
+    /**
+     * @var Node|null
+     */
+    private $elseNode;
+
+    public function enterNode(Node $node, Environment $env): Node
+    {
+        if ($this->isStructureNode($node)) {
```

Comment:
I think we need a `if ($node instanceof ModuleNode) {` that would reset `structureStack`.

Reply:
TBH; I'm kind of out of my league here as I'm not familiar with the ModuleNode.
I see I missed the `EmbedNode` and `AutoEscapeNode` in the list of "structure" nodes, not sure if or how this related to your suggestion though.

Reply:
SandboxNode and AutoEscapeNode could be allowed without impacting the breaking of loops, if we change the compiled code to use a `try/finally` to ensure that their cleanup logic also runs if a `break;` interrupts the execution of a parent loop from inside the node: https://3v4l.org/RotPq (we would have to check how this impacts error handling). This might apply to more tags relying on a cleanup phase (WithNode for instance).
However, such need means that the introduction of `break` in Twig is not a small change. It could be considered as a BC break because it is the first time that a syntax allows to have jumps in the execution of the AST (a deep descendant node impacts the execution path of a node much higher), which is a significant language change (other features have to know how to interact with it)

Blocks are macros are hard blockers here because they are compiled as separate PHP functions.
And the `{% embed %}` is parsed separately as a ModuleNode. It is not a child in the AST (EmbedNode represents the call to render this associated anonymous template)

Reply:
Thanks for the details. I see no way how to handle this so I'm closing the PR :+1: